cmd/controller/controller.go

```diff
@@ -0,0 +1,163 @@
+	NDMHostKey  = "kubernetes.io/hostname" // NDMHostKey is host name label prefix.
+	NDMActive   = "Active"                 // NDMActive is constant for active resource status.
+	NDMInactive = "Inactive"               // NDMInactive is constant for inactive resource status.
+)
+
+// ProbeChannel is used to send a copy of controller object to each probe.
+// Each probe can get the copy of controller struct any time they
+// need to read the channel.
+var ProbeChannel = make(chan *Controller)
+
+// Controller is the controller implementation for do resources
+type Controller struct {
+	HostName      string               // host name is host name in which disk is attached
+	KubeClientset kubernetes.Interface // kubeclientset is standard kubernetes clientset
+	Clientset     clientset.Interface  // sampleclientset is clientset for our own API group
+}
+
+// newController returns a controller pointer for any error case it will return nil
+func newController(kuberconfig string) (*Controller, error) {
+	masterURL := ""
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		if kuberconfig == "" {
+			return nil, errors.New("kubeconfig is empty")
+		}
+		cfg, err = clientcmd.BuildConfigFromFlags(masterURL, kuberconfig)
+		if err != nil {
+			return nil, err
+		}
+	}
+	kubeClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	crdClient, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	host, ret := os.LookupEnv("NODE_NAME")
+	if ret != true {
+		return nil, errors.New("error building hostname")
+	}
+	controller := &Controller{
+		HostName:      host,
+		KubeClientset: kubeClient,
+		Clientset:     crdClient,
+	}
+	return controller, nil
+}
+
+// Start scans the locally attached disks and push them to etcd.
+// This function is called when we execute cli command ndm start.
+func Start(kuberconfig string) {
+	// set up signals so we handle the first shutdown signal gracefully
+	stopCh := signals.SetupSignalHandler()
+	controller, err := newController(kuberconfig)
+	if err != nil {
+		glog.Fatal(err)
+	}
+	if err = controller.run(2, stopCh); err != nil {
+		glog.Fatalf("error running controller: %s", err.Error())
+	}
+}
+
+// run will discover the local disks and push them to the etcd.
+func (c *Controller) run(threadiness int, stopCh <-chan struct{}) error {
+	glog.Info("started the controller")
+	// sending controller object to each probe. Each probe can get a copy of
+	// controller struct anytime only they need to read channel.
+	go func() {
+		for true {
+			ProbeChannel <- c
+		}
+	}()
+	<-stopCh
+	glog.Info("shutting down the controller")
+	return nil
+}
+
+// DeviceList queries all the disk resources for this host from etcd and
+// prints them to the standard output. This function is called when we execute
+// cli command ndm device list.
+func DeviceList(kuberconfig string) {
+	controller, err := newController(kuberconfig)
```

Comment:
should this and the following functions be moved to diskstor?

Reply:
`diskstor.go` contains crate/update/delete/deactivate of disk .. is it ok to put `DeviceList()` there. as this function is used by cli command `ndm device list`